### PR TITLE
[CALCITE-3399] Field-pruning for set operators (except UNION ALL) changes query semantics (Jin Xing)

### PR DESCRIPTION
**Pull request by Aaaaaaron**
Pre discussion in #1495. The author is no response anymore, so I will promote this.

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java [hunk 1]
```diff
@@ -885,14 +886,22 @@ public TrimResult trimFields(
...
       ImmutableBitSet fieldsUsed,
       Set<RelDataTypeField> extraFields) {
     final RelDataType rowType = setOp.getRowType();
     final int fieldCount = rowType.getFieldCount();
+
+    // Trim fields only for UNION ALL.
+    // Operator UNION/INTERSECT/INTERSECT ALL/EXCEPT/EXCEPT ALL
+    // will check duplication. We can not trim fields on them.
+    if (!(setOp.kind == SqlKind.UNION && setOp.all)) {
+      return result(setOp, Mappings.createIdentity(fieldCount));
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: Does `INTERSECT ALL` and `EXCEPT ALL` check the duplicates ?

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: > Does `INTERSECT ALL` and `EXCEPT ALL` check the duplicates ?

Seems they don't. This is in the ori JIRA, should I remove them from the can not trim list?

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: You may need to check it with different db engines, here https://www.db-fiddle.com/

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: PG:
image

image

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: PG does have the difference, and MySQL seems don't support intersect all/except all.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: And what is your conclusion there ？

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: > And what is your conclusion there ？

We can not trim fields on the operator INTERSECT ALL/EXCEPT ALL

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: My test shows that INTERSECT/EXCEPT ALL does not distinct values, so we can trim fields for them.
Here is my test query:
```sql
create table a (
  a1 int,
  a2 int
);

insert into a values (1, 2);
insert into a values (3, 4);
insert into a values (3, 4);
insert into a values (1, 2);
insert into a values (5, 6);
insert into a values (5, 6);

create table b (
  b1 int,
  b2 int
);

insert into b values (1, 2);
insert into b values (3, 4);
insert into b values (3, 4);
```

```sql
select * from a except all select * from b;
```

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: INTERSECT/EXCEPT ALL does not distinct values, but we still can not trim fields for them.
Take the pic below, for example, q1 is the trimmed form of q2, and we can the result in the pic, they are different:
q1:
```
select name from
(select name from emp0
intersect all
select name from emp1)
```

q2:
```
select name from
(select name, deptno from emp0
intersect all
select name, deptno from emp1)
```

image

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql2rel/RelFieldTrimmer.java:902]: Hmm, i think i got your point: any set operations that have comparison between branches can not trim fields.

**Review by danny0405 — APPROVED** [review]
+1, please resolve the conflicts.

**Pull request closed without merging**